### PR TITLE
Implement local dependencies proposal

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
<rdar://problem/39418745> [SR-7433]: Implement SE-201 Package Manager Local Dependencies

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
This is mostly complete but needs actual tests and I need to make sure I didn't miss any edge case. @hartbit I'll appreciate if you can review the code

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

File: Sources/PackageDescription4/PackageDependency.swift [hunk 1]
```diff
@@ -24,6 +24,12 @@ extension Package.Dependency: Equatable {
         url: String,
         _ requirement: Package.Dependency.Requirement
     ) -> Package.Dependency {
+        // FIXME: This is suboptimal but its the only way to do this right now.
```
Comment (hartbit) [inline Sources/PackageDescription4/PackageDependency.swift]: Hmm, could we have made dependency into an enum:

```swift
public enum Dependency {
    case remote(RemoteDependency)
    case local(LocalDependency)
}

public class RemoteDependency {
    /// The dependency requirement.
    public enum Requirement {
        case exactItem(Version)
        case rangeItem(Range<Version>)
        case revisionItem(String)
        case branchItem(String)
    }

    /// The url of the dependency.
    public let url: String

    /// The dependency requirement.
    public let requirement: Requirement

    /// Create a dependency.
    init(url: String, requirement: Requirement) {
        self.url = url
        self.requirement = requirement
    }
}

public class LocalDependency {
    public let path: String
}
```

File: Sources/PackageGraph/DependencyResolver.swift [hunk 2]
```diff
@@ -17,6 +17,7 @@ public enum DependencyResolverError: Error, Equatable, CustomStringConvertible {
     case unsatisfiable
 
     /// The resolver encountered a versioned container which has a revision dependency.
+    // FIXME: Rename this to incompatible constraints.
```
Comment (hartbit) [inline Sources/PackageGraph/DependencyResolver.swift]: Didn't you want to rename it now?

File: Sources/PackageGraph/PackageGraphRoot.swift [hunk 3]
```diff
@@ -133,6 +134,9 @@ extension PackageDescription4.Package.Dependency.Requirement {
 
         case .exactItem(let version):
             return .versionSet(.exact(Version(pdVersion: version)))
+
+        case .localPackageItem:
```
Comment (hartbit) [inline Sources/PackageGraph/PackageGraphRoot.swift]: I'm still confused as to why we're representing local packages through `Requirement`. Is this the only solution?

File: Sources/Workspace/Workspace.swift [hunk 4]
```diff
@@ -400,9 +400,9 @@ extension Workspace {
         guard let dependency = diagnostics.wrap({ try managedDependencies.dependency(forNameOrIdentity: packageName) }) else {
             return
         }
-        guard case .checkout(let currentState) = dependency.state else {
-            let error = WorkspaceDiagnostics.DependencyAlreadyInEditMode(dependencyName: packageName)
-            return diagnostics.emit(error)
+        // FIXME: This should be "can resolve" and not "can edit".
```
[the remark this answers is not in the document]
Comment (aciidb0mb3r) [reply Sources/Workspace/Workspace.swift]: Nope, just a low-priority change.

File: Sources/PackageDescription4/PackageDependency.swift (hunk 1 above)
Comment (aciidb0mb3r) [reply Sources/PackageDescription4/PackageDependency.swift]: Possibly...I plan to revisit these data structure as part of upcoming runtime refactor.

Comment (aciidb0mb3r) [reply Sources/PackageDescription4/PackageDependency.swift]: Also note that we can't change the public facing data structures.

File: Sources/PackageGraph/PackageGraphRoot.swift (hunk 3 above)
Comment (aciidb0mb3r) [reply Sources/PackageGraph/PackageGraphRoot.swift]: Requirement is correct though as it *is* a requirement.

File: Sources/PackageDescription4/PackageDependency.swift (hunk 1 above)
Comment (hartbit) [reply Sources/PackageDescription4/PackageDependency.swift]: Can’t we change the public facing data structures as part of a 4.2 PD?
